file-iodp.8
.\" Automatically generated by Pod::Man version 1.04
.\" Thu Jan 18 18:45:33 2001
.\"
.\" Standard preamble:
.\" ======================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R

.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used
.\" to do unbreakable dashes and therefore won't be available.  \*(C` and
.\" \*(C' expand to `' in nroff, nothing in troff, for use with C<>
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` `
.    ds C' '
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr
.\" for titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and
.\" index entries marked with X<> in POD.  Of course, you'll have to process
.\" the output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
.    .
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it
.\" makes way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.bd B 3
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ======================================================================
.\"
.IX Title "FILE-IODP 1"
.TH FILE-IODP 1 "0.2.4" "2001-01-18" "iodp"
.UC
.SH "NAME"
file-iodp \- set \s-1IODP\s0 environment for file handles
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
file-iodp [options] filename program [arguments]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBfile-iodp\fR opens a file and runs
\&\fIprogram\fR with the remaining arguments. In an alternative 
operation mode \fBfile-iodp\fR may just set the \s-1IODP\s0 environment
variables for an existing file handle and run \fIprogram\fR.
.PP
\&\fIoptions\fR is a is a series of getopt or  getopt_long style options. See
below for more information.
.SH "OPTIONS"
.IX Header "OPTIONS"
.Sh "File options:"
.IX Subsection "File options:"
Note: The default is to open the file read-write, in non-truncating
overwrite mode (like O_RDWR does).
.Ip "\-r, \-\-rdonly" 8
.IX Item "-r, --rdonly"
Open the file read-only.
.Ip "\-w, \-\-wronly" 8
.IX Item "-w, --wronly"
Open the file write-only. 
.Ip "\-a, \-\-append" 8
.IX Item "-a, --append"
Append to end of file. Do not truncate.
.Ip "\-c, \-\-create" 8
.IX Item "-c, --create"
Create new file, if needed.
.Ip "\-e, \-\-exclusive" 8
.IX Item "-e, --exclusive"
Fail if file already exists.
.Ip "\-t, \-\-truncate" 8
.IX Item "-t, --truncate"
Truncate file.
.Ip "\-f, \-\-fifo" 8
.IX Item "-f, --fifo"
Create fifo, not file.
.Ip "\-m, \-\-mode=MODE" 8
.IX Item "-m, --mode=MODE"
Set access rights of the file to \s-1MODE\s0.
The default is 0644 for files and 622 for fifos.
.Ip "\-R, \-\-unlink" 8
.IX Item "-R, --unlink"
Remove existing file. This will be done before any other operation.
.Ip "\-n, \-\-reuse=FD" 8
.IX Item "-n, --reuse=FD"
Reuse existing file handle \s-1FD\s0, do not open another file. \fBfile-iodp\fR
will act as if it opened the file and got the file handle \fB\s-1FD\s0\fR in
return.
.Sh "Child program options:"
.IX Subsection "Child program options:"
.Ip "\-C, \-\-chroot=DIR" 8
.IX Item "-C, --chroot=DIR"
Change the working directory to \fI\s-1DIR\s0\fR. Then change the
root directory of the process to \fI.\fR (the current
directory).
This will be done after the socket has been created
and before changing the group or user \s-1ID\s0.
.Ip "\-F, \-\-fd=N" 8
.IX Item "-F, --fd=N"
Move socket handle to file descriptor \fIN\fR.
.Sp
\&\fIN\fR is an integer in the range from 0 to a system
"dependent limit. The default is to not move it.
Note that moving the socket to file descriptor 2
may disturb the output of error messages.
.Ip "\-g, \-\-gid=GID" 8
.IX Item "-g, --gid=GID"
Group \s-1ID\s0 to change to.
.Sp
tcpsocket will switch to that group id after
the socket has been set up. \s-1GID\s0 has to be a
positive integer.
.Ip "\-u, \-\-uid=UID" 8
.IX Item "-u, --uid=UID"
User \s-1ID\s0 to change to.
.Sp
tcpsocket will switch to that user id after
the socket has been set up. \s-1UID\s0 has to be a
positive integer.
.Ip "\-U, \-\-uidgid" 8
.IX Item "-U, --uidgid"
The same as \fB\-u\fR \f(CW$UID\fR \fB\-g\fR \f(CW$GID\fR.
.Sp
Usually used together with envuidgid.
.Sh "Data-gathering options:"
.IX Subsection "Data-gathering options:"
.Ip "\-y, \-\-symbolic-name=NAME" 8
.IX Item "-y, --symbolic-name=NAME"
Use \fI\s-1NAME\s0\fR as symbolic socket name.
.Sp
This is used for \f(CW$IODP_n_SYMNAME\fR, and may be
used by \s-1IODP\s0 compliant application in an
application dependant way. The default is to not
set it.
.Sh "Verbosity options:"
.IX Subsection "Verbosity options:"
.Ip "\-1" 8
.IX Item "-1"
Print the local port number to the standard output.
.Ip "\-q, \-\-quiet" 8
.IX Item "-q, --quiet"
Quiet. Do not print error messages.
.Ip "\-Q" 8
.IX Item "-Q"
Print error messages.
.Sp
This is the default.
.Ip "\-v, \-\-verbose" 8
.IX Item "-v, --verbose"
Verbose. Print error and status messages.
.Ip "\*(--version" 8
.IX Item "version"
Show version information.
.Ip "\*(--help" 8
.IX Item "help"
Show a list of options or the long help on one.
.Sp
The use with an argument shows the long helptext
of that option, without an argument it will list
all options.
.Ip "\*(--longhelp" 8
.IX Item "longhelp"
Show longer help texts for all or one variable
.Sp
The use with an argument shows the long help for
this option, without arguments it shows the long
description of all options.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
\&\fBtcpsocket\fR sets the environment variables documented in the
\&\fIiodp-tcp\fR specification.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.Vb 2
\&        file-iodp -r /proc/kmsg \e
\&        /usr/bin/logger -t kernel
.Ve
.SH "BUGS"
.IX Header "BUGS"
unknown.
.SH "AUTHOR"
.IX Header "AUTHOR"
Uwe Ohse
.SH "SEE ALSO"
.IX Header "SEE ALSO"
tcpsocket(8), udpsocket(8), uxstreamsocket(8), uxdgramsocket(8),
iodp(5).



NEWS
Version 0.2.4, January 2001

* added sample echo server.


file-iodp.c
#include "_stralloc.h"
#include "error.h"
#include "uogetopt.h"
#include "_alloc.h"
#include "fd.h"
#include "str2num.h"
#include "byte.h"
#include "fmt.h"
#include "bailout.h"
#include "pathexec.h"
#include "iodptools.h"
#include "fifo.h"
#include "_str.h"
#include "seek.h"
#include <fcntl.h>
#include <unistd.h>
#include <sys/stat.h>

#define PROGNAME "file-iodp"
const char *progname=PROGNAME;

static const char *opt_mode;
const char *opt_symname;
static int opt_unlink;
static int opt_rdwr=O_RDWR;
static int opt_append=0;
static int opt_create=0;
static int opt_excl=0;
static int opt_trunc=0;
static int opt_fifo=0;
static long opt_reuse=-1;

uogetopt_t localopts_socket[] = {
{0,"",UOGO_TEXT, 0,0, "File options:\n",0},
{'r',"rdonly",UOGO_FLAG,&opt_rdwr,O_RDONLY,
"Open the file read-only.",0},
{'w',"wronly",UOGO_FLAG,&opt_rdwr,O_WRONLY,
"Open the file write-only.",0},
{'a',"append",UOGO_FLAG,&opt_append,O_APPEND,
"Append to end of file.\n"
"Do not truncate.",0},
{'c',"create",UOGO_FLAG,&opt_create,O_CREAT,
"Create new file, if needed.",0},
{'e',"exclusive",UOGO_FLAG,&opt_excl,O_EXCL,
"Fail if file already exists.",0},
{'t',"truncate",UOGO_FLAG,&opt_trunc,O_TRUNC,
"Truncate file.",0},
{'f',"fifo",UOGO_FLAG,&opt_fifo,1,
"Create fifo, not file.",0},
{'m', "mode", UOGO_STRING, &opt_mode, 0,
"Set access rights of the file to MODE.\n"
"The default is 0644 for files and 622 for fifos.", "MODE"},
{'n',"reuse",UOGO_LONG,&opt_reuse,1,
"Reuse existing file handle FD.\n"
"Do not open another file","FD"},
{'R',"unlink",UOGO_FLAG,&opt_unlink,1,
"Remove existing file.",0},
{0, 0, 0, 0, 0, 0, 0}
};


uogetopt_t localopts_data[] = {
{'y', "symbolic-name", UOGO_STRING, &opt_symname, 0,
"Use NAME as symbolic file name.\n"
"This is used for $IODP_n_LOCALNAME, and may be\n"
"used by IODP compliant application in an\n"
"application dependant way. The default is to not\n"
"set it.","NAME"},
{0, 0, 0, 0, 0, 0, 0}
};

uogetopt_t localopts_verbose[] =
{
{0, 0, 0, 0, 0, 0, 0}
};

static const char *path;
const char *usagestr=
"usage: " PROGNAME " filename program [args]";
int skipargs=2;

int doopen(char **argv)
{
	int fd;
	int mode;
	unsigned long acmode;
	path=argv[1];
	if (!opt_mode) {
		if (opt_fifo)
			opt_mode="0622";
		else
			opt_mode="0644";
	}
	if (opt_reuse != -1)
		return opt_reuse;
	if (str2ulong(opt_mode,&acmode,8)!=(int) str_len(opt_mode))
		xbailout(2,0,"can't understand access mode ",opt_mode,0,0);
	if (opt_unlink)
		if (-1==unlink(path) && errno!=error_noent)
			bailout(errno,"cannot unlink ",path,0,0);
	if (opt_fifo) {
		int ret;
		ret=fifo_make(path,acmode);
		if (ret==-1) bailout(errno,"cannot open fifo ",path,0,0);
	}

	mode=opt_rdwr | opt_append | opt_create | opt_trunc | opt_excl 
		| O_NONBLOCK;
	fd=open(path,mode,acmode);
	if (fd==-1) bailout(errno,"cannot open ",path,0,0);

	if (-1==fchmod(fd,acmode))
		bailout(errno,"fchmod ",path,0,0);
	return fd;
}
void doiodp(int fd)
{
	const char *m;
	stralloc ma=STRALLOC_INIT;
	off_t x,y;
	if (opt_fifo)
		m="rw";
	else if (opt_rdwr==O_RDONLY)
		m="r";
	else if (opt_rdwr==O_WRONLY)
		m="w";
	else
		m="rw";
	if (!stralloc_copys(&ma,m)) oom();
	if (isatty(fd)) {
		if (!stralloc_cats(&ma,"T")) oom();
	}
	else {
		x=seek_cur(fd);
		if (x!=-1) {
			y=seek_begin(fd);
			if (y!=-1) {
				y=seek_set(fd,x);
				if (y==-1) xbailout(111,errno,"lseek",0,0,0);
				if (!stralloc_cats(&ma,"S")) oom();
			}
		}
	}
	if (!stralloc_0(&ma)) oom();

	iodp_set(fd,"FILE",0,0,path,opt_symname,ma.s);
}

